control-center/include/etc/confluent/docker/control-center.properties.template

```diff
@@ -1,54 +1,67 @@
+{% set required_props = {
+  'bootstrap.servers': ['CONTROL_CENTER_BOOTSTRAP_SERVERS'],
+  'zookeeper.connect': ['CONTROL_CENTER_ZOOKEEPER_CONNECT'],
+  'confluent.controlcenter.data.dir': ['CONTROL_CENTER_DATA_DIR'],
+  'confluent.monitoring.interceptor.topic.replication': ['CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_REPLICATION', 'CONTROL_CENTER_REPLICATION_FACTOR'],
+  'confluent.controlcenter.internal.topics.replication': ['CONTROL_CENTER_INTERNAL_TOPICS_REPLICATION', 'CONTROL_CENTER_REPLICATION_FACTOR'],
+  'confluent.controlcenter.command.topic.replication': ['CONTROL_CENTER_COMMAND_TOPIC_REPLICATION', 'CONTROL_CENTER_REPLICATION_FACTOR'],
+  'confluent.metrics.topic.replication': ['CONTROL_CENTER_METRICS_TOPIC_REPLICATION', 'CONFLUENT_METRICS_TOPIC_REPLICATION', 'CONTROL_CENTER_REPLICATION_FACTOR']
+} -%}
 
-{# required properties #}
-bootstrap.servers={{env['CONTROL_CENTER_BOOTSTRAP_SERVERS']}}
-zookeeper.connect={{env['CONTROL_CENTER_ZOOKEEPER_CONNECT']}}
-confluent.controlcenter.data.dir={{env['CONTROL_CENTER_DATA_DIR']}}
-confluent.monitoring.interceptor.topic.replication={{env.get('CONTROL_CENTER_MONITORING_INTERCEPTOR_TOPIC_REPLICATION', env['CONTROL_CENTER_REPLICATION_FACTOR'])}}
-confluent.controlcenter.internal.topics.replication={{env.get('CONTROL_CENTER_INTERNAL_TOPICS_REPLICATION', env['CONTROL_CENTER_REPLICATION_FACTOR'])}}
-confluent.controlcenter.command.topic.replication={{env.get('CONTROL_CENTER_COMMAND_TOPIC_REPLICATION', env['CONTROL_CENTER_REPLICATION_FACTOR'])}}
-confluent.metrics.topic.replication={{env.get('CONTROL_CENTER_METRICS_TOPIC_REPLICATION', env['CONTROL_CENTER_REPLICATION_FACTOR'])}}
+{% set metrics_props = {
+    'confluent.metrics.topic': ['CONTROL_CENTER_METRICS_TOPIC', 'CONFLUENT_METRICS_TOPIC'],
+    'confluent.metrics.topic.retention.ms': ['CONTROL_CENTER_METRICS_TOPIC_RETENTION_MS', 'CONFLUENT_METRICS_TOPIC_RETENTION_MS'],
+    'confluent.metrics.topic.partitions': ['CONTROL_CENTER_METRICS_TOPIC_PARTITIONS', 'CONFLUENT_METRICS_TOPIC_PARTITIONS'],
+    'confluent.metrics.topic.skip.backlog.minutes': ['CONTROL_CENTER_METRICS_TOPIC_SKIP_BACKLOG_MINUTES', 'CONFLUENT_METRICS_TOPIC_SKIP_BACKLOG_MINUTES']
+} -%}
 
-{# optional properties #}
-{% set other_props = {
-    'CONTROL_CENTER_ID': 'confluent.controlcenter.id',
```

Comment:
Any reason these props are not explicitly listed in the new template?

Reply:
It's because below code from line 60-67 can cover them:
```
{% set excludes = [] -%}
{{ SET_PROPERTIES(required_props, true, excludes) }}
{{ SET_PROPERTIES(metrics_props, false, excludes) }}
{{ SET_PROPERTIES(monitoring_props, false, excludes) }}
{{ SET_PROPERTIES(other_props, false, excludes) }}
{{ SET_PROPERTIES_WITH_ENV_TO_PROPS('CONTROL_CENTER_', 'confluent.controlcenter.', excludes) }}
{{ SET_PROPERTIES_WITH_ENV_TO_PROPS('CONFLUENT_METADATA_', 'confluent.metadata.', excludes) }}
{{ SET_PROPERTIES_WITH_ENV_TO_PROPS('CONFLUENT_SUPPORT_', 'confluent.support.', excludes) }}
```

Basically the new logic will first set the required properties, then set optional metrics, monitoring, other properties (other properties are special ones that don't have common prefix). Then, the logic will set ones that have common prefixes, for example `CONTROL_CENTER_*`. 

For example, `CONTROL_CENTER_ID` and `CONTROL_CENTER_NAME` have a regular prefix and they would be set by line 65
```
{{ SET_PROPERTIES_WITH_ENV_TO_PROPS('CONTROL_CENTER_', 'confluent.controlcenter.', excludes) }}
```


One thing to note is that `CONTROL_CENTER_MONITORING_*` couldn't be set using `SET_PROPERTIES_WITH_ENV_TO_PROPS` (extracting common prefix) is because they start with prefix `CONTROL_CENTER_`, so later in line 65, they will be falsely translated to `confluent.controlcenter.monitoring.*` again. Therefore `SET_PROPERTIES` is used, where it not only translates the properties but also adds the properties to `excludes` list so that line 65 wouldn't translate them again.

Same for `CONTROL_CENTER_METRICS_*`, but it's further complicated by the precedence.

Reply:
`SET_PROPERTIES` macro 
- handles special translation (e.x. `CONTROL_CENTER_MONITORING_` to `confluent.monitoring.`)
- handles precedence (e.x. `CONTROL_CENTER_METRICS_` is searched before `CONFLUENT_METRICS_`)
- handles required vs. not required properties (e.x. if no environment variable is set, required properties are still set with empty string but not required properties are not set at all)

`SET_PROPERTIES_WITH_ENV_TO_PROPS` macro
- handles special translation
- only handles not required properties

Reply:
Makes sense. I misread it, assumed id and name were  in required set.